pkg/domain/infosync/info.go

```diff
@@ -501,13 +510,57 @@ func doRequest(ctx context.Context, apiName string, addrs []string, route, metho
 				return nil, err
 			}
 			if res.StatusCode != http.StatusOK {
-				logutil.BgLogger().Warn("response not 200",
+				var (
+					leaderAddr string
+				)
+				for _, addr2 := range addrs {
+					status, err2 := etcdCli.Status(ctx, addr2)
+					if err2 != nil {
+						logutil.BgLogger().Warn(
+							"failed to get leader by checking endpoint status",
+							zap.String("endpoint", addr2),
+							zap.Error(err2))
+						continue
+					}
+					if status.Leader == status.Header.MemberId {
+						leaderAddr = addr2
+						break
+					}
+				}
+
+				logutil.BgLogger().Warn("response not 200, will retry sending to etcd leader",
 					zap.String("method", method),
 					zap.String("hosts", addr),
 					zap.String("url", url),
 					zap.Int("http status", res.StatusCode),
 					zap.Int("address order", idx),
+					zap.String("leader", leaderAddr),
 				)
+				if leaderAddr != "" {
```

Comment:
If `leaderAddr` is the same as the requested address, should we still make the request?

Reply:
https://github.com/pingcap/tidb/pull/56332/commits/1e74686074b7bb90017bed7f4c93cec43ffadb98